chapter_01/final_exercise_01.md

```diff
@@ -21,66 +21,129 @@
+
 23. **Q:** What is HiveQL?
+23. **A:** Hiveql is a sql like query language that can be used to query on the data in hive.
+
 24. **Q:** What are partitions in Hive?
+24. **A:** Partitions are parts of the table split between machines used in case the tables are to big in order to distribute the processing and the data which in return speeds up the process of analyzing and querying on the data and for less data to be processed using pruning.
+
 25. **Q:** Can you define SerDe in the context of Hive?
+25. **A:** SerDe is the Serializer and the Deserializer which decides how hive stores and reads the data.
 
 ### Chapter 6: Apache ZooKeeper
 
 26. **Q:** What is Apache ZooKeeper and what role does it play in a distributed environment?
+26. **A:** Zookeeper is a distributed configuration manager  and coordinator that holds different configurations for different systems which allows different systems to coordinate with each other and allows for a single source of configuration change.
+
 27. **Q:** Can you explain the concept of znodes in ZooKeeper?
+27. **A:** ZNodes are Like data nodes which can have children, which in turn acts like files and directories.
+
 28. **Q:** How does ZooKeeper handle coordination and configuration management across distributed systems?
+28. **A:** Zookeeper handles coordination and configuration management using watches where clients can listen to changes, and atomicity and sessions which aid in the process of the coordination
+
 29. **Q:** What are ephemeral nodes and sequence nodes in ZooKeeper, and how are they used?
+29. **A:** *ephermal nodes* are nodes which last for as long as the session is kept alive which can be used for master / leader allocation and *sequence nodes* are nodes which have an assigned unique number, can be used for distributed counters.
+
 30. **Q:** How does ZooKeeper ensure data consistency and reliability across distributed nodes?
+30. **A:** Zookeeper ensures consistency and reliability by replication and only the primary node being the one interacted with, in adition to atomicity.
 
 ### Chapter 7: Apache HBase
 
 31. **Q:** What is Apache HBase?
 32. **Q:** What is a Row Key in HBase?
 33. **Q:** What is a Column Family in HBase?
-34. **Q:** How does HBase ensure data availability and fault tolerance?
+34. **Q:** How oes HBase ensure data availability and fault tolerance?
 35. **Q:** What is the role of ZooKeeper in an HBase environment?
 
 ### Chapter 8: Apache Kafka
 
 36. **Q:** What is Apache Kafka?
+36. **A:** Kafka is a distributed, replicated Streaming pub sub model which allows for event driven and event streaming architecture.
+
 37. **Q:** What are Topics in Kafka?
+37. **A:** Topics in kafka are a way to organize data, a topic is like a queue of data (could be multiple queues for distributed purposes and for performance improvements) which we can subscribe or publish to.
```

Comment:
So is it a topic or a queue? Explain the difference between a topic and a queue. What does it consists of?